pkg/util/vsphere/vm/power_state.go

```diff
@@ -370,35 +400,341 @@ func doAndWaitOnSoftPowerOp(
+		ctx,
+		client,
+		mo,
+		fetchProperties,
+		desiredLastRestartTime,
+		powerOpBehavior)
+}
+
+const (
+	propConfigExtraConfig = "config.extraConfig"
+
+	// ExtraConfigKeyLastRestartTime is the name of the key in a VM's
+	// ExtraConfig array that contains the epoch of the last time the VM was
+	// restarted.
+	ExtraConfigKeyLastRestartTime = "vmservice.lastRestartTime"
+)
+
+func restart(
+	ctx context.Context,
+	client *vim25.Client,
+	mo mo.VirtualMachine,
+	fetchProperties bool,
+	desiredLastRestartTime time.Time,
+	powerOpBehavior PowerOpBehavior) (PowerOpResult, error) {
+
+	var (
+		err                          error
+		currentPowerState            types.VirtualMachinePowerState
+		currentLastRestartTime       *time.Time
+		currentLastRestartTimeString string
+		propsToFetch                 []string
+		result                       PowerOpResult
+		log                          = logr.FromContextOrDiscard(ctx)
+		obj                          = object.NewVirtualMachine(client, mo.Self)
+		desiredLastRestartTimeString = desiredLastRestartTime.Format(time.RFC3339)
+	)
+
+	// Fetch the VM's current power state and extra config if they are not
+	// already known or explicitly requested.
+	if fetchProperties || mo.Summary.Runtime.PowerState == "" {
+		propsToFetch = append(propsToFetch, object.PropRuntimePowerState)
+	}
+	if fetchProperties || mo.Config == nil || len(mo.Config.ExtraConfig) == 0 {
+		propsToFetch = append(propsToFetch, propConfigExtraConfig)
+	}
+	if len(propsToFetch) > 0 {
+		if err := obj.Properties(
+			ctx,
+			mo.Self,
+			propsToFetch,
+			&mo); err != nil {
+			return 0, fmt.Errorf("failed to retrieve properties %w", err)
+		}
+	}
+	if mo.Config == nil {
+		return 0, errors.New("nil vm.config")
+	}
+	currentPowerState = mo.Summary.Runtime.PowerState
+
+	if currentPowerState != types.VirtualMachinePowerStatePoweredOn {
+		log.V(5).Info(
+			"cannot restart VM in current power state",
+			"currentPowerState", currentPowerState,
+			"desiredLastRestartTime", desiredLastRestartTimeString,
+			"desiredPowerOpBehavior", powerOpBehavior.String())
+		return PowerOpResultNone, nil
+	}
+
+	// Find the currentLastRestartTime as recorded in the extra config.
+	currentLastRestartTime, err = GetLastRestartTimeFromExtraConfig(mo.Config.ExtraConfig)
+	if err != nil {
+		return 0, err
+	}
+	if currentLastRestartTime != nil {
+		currentLastRestartTimeString = currentLastRestartTime.Format(time.RFC3339)
+
+		// Do not restart if there is a recorded, last restart time that occurs
+		// after the desired last restart time.
+		if currentLastRestartTime.After(desiredLastRestartTime) {
+			log.V(5).Info(
+				"will not restart VM as desiredLastRestartTime < currentLastRestartTime",
+				"currentPowerState", currentPowerState,
+				"currentLastRestartTime", currentLastRestartTimeString,
+				"desiredLastRestartTime", desiredLastRestartTimeString,
+				"desiredPowerOpBehavior", powerOpBehavior.String())
+			return PowerOpResultNone, nil
```

Comment:
I understand that an error isn't returned here to avoid unnecessary reconciling. However, this means that users won't know if they specify `vm.spec.lastRestart` before `currentLastRestartTime`. I wonder if we could submit a K8s event or have a `vm.status.lastRestart` field to display this information for a better user experience?

Reply:
> vm.status.lastRestart

I was going to add this prior to requesting a review, so good catch! :)